src/Concretize.hs

```diff
@@ -136,6 +136,14 @@ concretizeXObj allowAmbiguityRoot typeEnv rootEnv visitedDefinitions root =
                 SymPath spath name = rootDefinitionPath
                 lambdaPath = SymPath spath ("_Lambda_" ++ lambdaToCName name (envFunctionNestingLevel envWithArgs) ++ "_" ++ show (infoIdentifier ii) ++ "_env")
                 lambdaNameSymbol = XObj (Sym lambdaPath Symbol) (Just dummyInfo) Nothing
+                -- Anonymous functions bound to a let name might call themselves. These recursive instances will have already been qualified as LookupRecursive symbols.
+                -- Rename the recursive calls according to the generated lambda name so that we can call these correctly from C.
+                renameRecursives x = case x of
+                                       (XObj (Sym _ LookupRecursive) si st) -> (XObj (Sym lambdaPath LookupRecursive) si st)
+                                       y@(XObj (Lst _) _ _) -> endoMap renameRecursives y
+                                       y@(XObj (Arr _) _ _) -> endoMap renameRecursives y
+                                       _ -> x
```

Comment:
I feel like this might be readable and more consistent with other parts of the code base if it pattern matched on the argument instead of using `case`?